src/librustc/ty/instance.rs

```diff
@@ -45,6 +45,16 @@ pub enum InstanceDef<'tcx> {
     CloneShim(DefId, Ty<'tcx>),
 }
 
+impl<'a, 'tcx> Instance<'tcx> {
+    pub fn instance_ty(&self,
```

Comment:
I'd call this `ty`. One thing I'm worried about is `trans_apply_param_substs` failing if this is accidentally used somewhere where a `ParamEnv` is needed or some such.
Shouldn't `self.def.def_ty(tcx).subst(tcx, self.substs)` suffice?

Reply:
@eddyb I don't really know the details of substitutions yet. I have updated this function to use `self.def.def_ty(tcx).subst(tcx, self.substs)` but then I get an ICE in `run_cargo` (see latest build error). Everything seems to work when I use `trans_apply_param_substs`. 

Should I still avoid `trans_apply_param_substs`? I can investigate further why it crashes if you want me to.